### PR TITLE
feat: add intial mechanism for billing options(admin)

### LINKED ISSUE #3126: Admin: Add "Billing" Subtab in Admin Settings
Invoices for events (not for ticket buyers!) need to have an Invoice Address of the Eventyay Platform provider (FOSSASIA). We can keep this information in the settings under the subtab "Billing". Please:
1. Add the subtab billing next to the tab "ticket fees" (https://eventyay.com/admin/settings/ticket-fees)
2. Create a heading "Invoice Handling" and provide fields for:
  * Invoice sent out - Dropdown - : First Day of Month, Second Day of Month, Third Day of Month
  * Timezone for sent outs: [Timezone here]
  * Invoice Format [Format Options here] (please resarch how to do it)
3. Create a heading "Invoice Contact" and provide fields for: 
  * Logo (enable image upload here), 
  * Contact Name, 
  * Phone, 
  * Email*, 
  * Country*, 
  * Tax ID, 
  * Registered Entity or Company,
  * Business Address*, 
  * City*, 
  * ZIP Code*, 
  * Additional Info, 
  * Comments

Screenshot from 2019-06-12 10-46-57

### PR DESCRIPTION
**Pull request by mrsaicharan1**
Screenshot 2019-06-17 at 4 00 59 PM
Added initial billing info subtab in admin

Added Invoice Contact info and Billing sub tab 

Fixes #3126
Server PR: https://github.com/fossasia/open-event-server/pull/6130

#### Short description of what this resolves:
Adds the front end part for configuring invoice settings such as date of invoice sent, timezone and invoice contact. 
Will be adding the invoice structure options in a follow-up PR.
#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mrsaicharan1** [comment]
@niranjan94 @CosmicCoder96 Where should the Invoice Contact billing info? I wasn't able to find a suitable model where I could store this.

**Comment by mariobehling** [comment]
@mrsaicharan1 Please always provide screenshots. Travis is failing currently.

**Comment by mrsaicharan1** [comment]
@mariobehling I've put up the screenshot.

**Comment by mariobehling** [comment]
@mrsaicharan1 

> - Invoice sent out - Dropdown - : First Day of Month, Second Day of Month, Third Day of Month
> - Timezone for sent outs: [Timezone here]
> - Invoice Format [Format Options here] (please resarch how to do it)

It should be like that in the UI:
Invoice Time to Sent Out: [ Dropdown here]
Timezone for Sent Outs: [ Dropdown here]
Invoice Format: [Edit Field here]
* [help text here about what options are possible in the invoice format, e.g. date, year, number, text..]

**Comment by mrsaicharan1** [comment]
> @mrsaicharan1
> 
> > * Invoice sent out - Dropdown - : First Day of Month, Second Day of Month, Third Day of Month
> > * Timezone for sent outs: [Timezone here]
> > * Invoice Format [Format Options here] (please resarch how to do it)
> 
> It should be like that in the UI:
> Invoice Time to Sent Out: [ Dropdown here]
> Timezone for Sent Outs: [ Dropdown here]
> Invoice Format: [Edit Field here] * [help text here about what options are possible in the invoice format, e.g. date, year, number, text..]


@mariobehling Yeah, got it. Storing different invoice formats and providing options for it, I was thinking about dealing it in a follow up PR as it would become really confusing to add everything in one PR.

**Comment by mariobehling** [comment]
Yes, you can deal with things step by step, but follow the structure and design, please. Thanks.

**Comment by mrsaicharan1** [comment]
I'll shift the admin invoice fields to the `settings` model/schema so that is more accessible as there exists only one record.

File: app/controllers/admin/settings/billing.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import Controller from '@ember/controller';
+import { timezones } from 'open-event-frontend/utils/dictionary/date-time';
+
+export default Controller.extend({
+  timezones,
+  updateModel() {
```
Comment (shreyanshdwivedi) [inline app/controllers/admin/settings/billing.js]: @mrsaicharan1 I think you can use async-await here. @CosmicCoder96 am I right?

Comment (mrsaicharan1) [reply app/controllers/admin/settings/billing.js]: @shreyanshdwivedi Yup.

File: app/controllers/admin/settings/billing.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    this.set('isLoading', true);
+    let adminInvoiceModel = this.model;
+    adminInvoiceModel.save()
+      .then(() => {
```
Comment (uds5501) [inline app/controllers/admin/settings/billing.js]: @shreyanshdwivedi in addition to that, I think this will be modified accordingly

Comment (shreyanshdwivedi) [reply app/controllers/admin/settings/billing.js]: Yep

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @uds5501 @shreyanshdwivedi  If we move forward with this, we can continue with other sub-parts of this payment invoice issue

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 not yet, will update it!

**Comment by CosmicCoder96** [comment]
@mrsaicharan1 okay, this can only be merged after server-side changes are done.

**Comment by mrsaicharan1** [comment]
Yeah sure.

**Comment by niranjan94** [comment]
@mrsaicharan1 do not open frontend PRs without the server implementation merged.

**Comment by mrsaicharan1** [comment]
Yeah will take care

**Comment by niranjan94** [comment]
@mrsaicharan1 closing this PR to prevent accidental merges. Please re-open once it is no longer blocked by server.

**Comment by mrsaicharan1** [comment]
@niranjan94 Can you please re-open this. The server-side changes for this have been merged.

**Comment by mrsaicharan1** [comment]
https://github.com/fossasia/open-event-server/pull/6098

**Comment by mrsaicharan1** [comment]
Would like to finalize this. Will need to make changes corresponding to the naming convention on the server side so that the data goes through.

**Comment by mrsaicharan1** [comment]
@niranjan94 Would like to finalize this. Will need to make changes corresponding to the naming convention on the server side so that the data goes through.

**Comment by mrsaicharan1** [comment]
The server side naming conventions have changed. Will update here too.

File: app/routes/admin/settings/billing.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import Route from '@ember/routing/route';
+
+export default Route.extend({
+  model() {
+    return this.store.findAll('setting');
```
Comment (CosmicCoder96) [inline app/routes/admin/settings/billing.js]: this is already called in the index route.

File: app/templates/admin/settings/billing.hbs [hunk 4]
```diff
@@ -0,0 +1,3 @@
+<div class="ui basic {{if isLoading 'loading' ''}} segment">
```
Comment (CosmicCoder96) [inline app/templates/admin/settings/billing.hbs]: no need for empty quotes.

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+  <h3 class="ui header">
+    {{t 'Invoice Handling'}}
+  </h3>
+  <div class="two fields">
+    {{#ui-dropdown class="selection" onChange=(action (mut model.invoiceSendingDay))}}
+      {{input type='hidden' id='invoiceSendingDay' value=model.invoiceSendingDay}}
+      <i class="dropdown icon"></i>
+      <div class="default text">{{t 'Invoices Sent Out'}}</div>
+      <div class="menu">
+        <div class="item" data-value="1">First of every month</div>
```
Comment (CosmicCoder96) [inline app/templates/components/forms/admin/settings/billing.hbs]: Translation helpers.

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        <div class="item" data-value="2">Second of every month</div>
```
Comment (CosmicCoder96) [inline app/templates/components/forms/admin/settings/billing.hbs]: Translation helpers.

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        <div class="item" data-value="3">Third of every month</div>
```
Comment (CosmicCoder96) [inline app/templates/components/forms/admin/settings/billing.hbs]: Translation helpers.

File: app/controllers/admin/settings/billing.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
+import Controller from '@ember/controller';
+import { timezones } from 'open-event-frontend/utils/dictionary/date-time';
+import { computed } from '@ember/object';
+
+export default Controller.extend({
+  timezones,
+  adminBillingEmail: computed('authManager.currentUser', function() {
```
Comment (CosmicCoder96) [inline app/controllers/admin/settings/billing.js]: why are you using a computed property for this?

Comment (mrsaicharan1) [reply app/controllers/admin/settings/billing.js]: @CosmicCoder96 When I use it without a computed property, it shows me that it's undefined `undefined is not an object (evaluating '(void 0).authManager')`.

File: app/routes/admin/settings/billing.js (hunk 3 above)
Comment (mrsaicharan1) [reply app/routes/admin/settings/billing.js]: Removed the redundant store hit & used `modelFor`

File: app/controllers/admin/settings/billing.js (hunk 8 above)
Comment (CosmicCoder96) [reply app/controllers/admin/settings/billing.js]: in the server schema, didn't you specify a separate field mail for admin billing? Why do you need this one?

File: app/components/forms/admin/settings/billing.js [hunk 9]
```diff
@@ -0,0 +1,124 @@
+import Component from '@ember/component';
+import FormMixin from 'open-event-frontend/mixins/form';
+
+export default Component.extend(FormMixin, {
```
Comment (niranjan94) [inline app/components/forms/admin/settings/billing.js]: Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

File: app/controllers/admin/settings/billing.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+import Controller from '@ember/controller';
+import { timezones } from 'open-event-frontend/utils/dictionary/date-time';
+import { computed } from '@ember/object';
+
+export default Controller.extend({
```
Comment (niranjan94) [inline app/controllers/admin/settings/billing.js]: Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

File: app/routes/admin/settings/billing.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+import Route from '@ember/routing/route';
+
+export default Route.extend({
```
Comment (niranjan94) [inline app/routes/admin/settings/billing.js]: Lets start moving to ES6 style classes for all new resources. 
https://github.com/fossasia/open-event-frontend/pull/3169/files

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 @niranjan94 Please review. Used ES6 syntax and made the requested changes.

File: app/templates/admin/settings/billing.hbs [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<div class="ui basic {{if isLoading 'loading'}} segment">
+  {{forms/admin/settings/billing timezones=timezones  adminUser=adminUser model=model save='updateInvoiceModel'}}
```
Comment (niranjan94) [inline app/templates/admin/settings/billing.hbs]: change `save='updateInvoiceModel'` to use closure actions

File: app/components/forms/admin/settings/billing.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        }
+      }
+    };
+
+  }
+
+  @action
+  submit() {
+    this.onValid(() => {
+      this.sendAction('save');
```
Comment (niranjan94) [inline app/components/forms/admin/settings/billing.js]: Do not use send action.

Comment (mrsaicharan1) [reply app/components/forms/admin/settings/billing.js]: @niranjan94 Used `this.attrs.save()` to invoke the action closure.

File: app/templates/admin/settings/billing.hbs (hunk 12 above)
Comment (mrsaicharan1) [reply app/templates/admin/settings/billing.hbs]: @niranjan94 Done!

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    <label class="required" for="adminBillingState">{{t 'State'}}</label>
+    {{input type='text' id='adminBillingState' value=model.adminBillingState}}
+  </div>
+  <div class="field">
+    <label class="required" for="adminBillingZip">{{t 'Zip Code'}}</label>
+    {{input type='text' id='adminBillingZip' value=model.adminBillingZip}}
+  </div>
+  <div class="field">
+    <label for="adminBillingAdditionalInfo">{{t 'Additional Info'}}</label>
+    {{input type='text' id='adminBillingAdditionalInfo' value=adminBillingAdditionalInfo}}
```
Comment (CosmicCoder96) [inline app/templates/components/forms/admin/settings/billing.hbs]: fix this.

Comment (mrsaicharan1) [reply app/templates/components/forms/admin/settings/billing.hbs]: @CosmicCoder96 Took care of this

File: app/components/forms/admin/settings/billing.js [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+        }
+      }
+    };
+
+  }
+
+  @action
+  submit() {
+    this.onValid(() => {
+      this.attrs.save();
```
Comment (niranjan94) [inline app/components/forms/admin/settings/billing.js]: Why attrs ? `this.save()` would work no ?

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 16]
```diff
@@ -0,0 +1,77 @@
+<form class="ui form" {{action 'submit' on='submit'}}>
```
Comment (uds5501) [inline app/templates/components/forms/admin/settings/billing.hbs]: Shouldn't there be an `ui form {{if isLoading 'loading'}}` for a loading UI?

Comment (mrsaicharan1) [reply app/templates/components/forms/admin/settings/billing.hbs]: It's there in `app/templates/admin/settings/billing.hbs`

Comment (uds5501) [reply app/templates/components/forms/admin/settings/billing.hbs]: @mrsaicharan1 Oh! alright, missed it.

File: app/components/forms/admin/settings/billing.js (hunk 15 above)
Comment (mrsaicharan1) [reply app/components/forms/admin/settings/billing.js]: @niranjan94 Yeah. It does!

File: app/components/forms/admin/settings/billing.js [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+          rules      : [
+            {
+              type   : 'empty',
+              prompt : this.l10n.t('Please enter additional info')
+            }
+          ]
+        }
+      }
+    };
+
```
Comment (Anupam-dagar) [inline app/components/forms/admin/settings/billing.js]: can you remove this empty line, this doesn't looks required.

**Comment by mrsaicharan1** [comment]
@niranjan94 @CosmicCoder96 Please review this as this is required to test the event invoice payment functionality as a whole.

File: app/controllers/admin/settings/billing.js [hunk 18]
```diff
@@ -0,0 +1,19 @@
+import Controller from '@ember/controller';
+import { timezones } from 'open-event-frontend/utils/dictionary/date-time';
+import { action } from '@ember/object';
+
+export default class extends Controller {
+  timezones = timezones;
+  @action
+  async updateInvoiceModel() {
+    this.set('isLoading', true);
+    let adminInvoiceModel = this.model;
```
Comment (niranjan94) [inline app/controllers/admin/settings/billing.js]: why store to a variable ?

Comment (mrsaicharan1) [reply app/controllers/admin/settings/billing.js]: @niranjan94 Had overlooked this. Update it now.

File: app/components/forms/admin/settings/billing.js [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+              type   : 'empty',
+              prompt : this.l10n.t('Please enter the Contact name')
+            }
+          ]
+        },
+        adminBillingPhone: {
+          identifier : 'adminBillingPhone',
+          rules      : [
+            {
+              type   : 'empty',
```
Comment (kushthedude) [inline app/components/forms/admin/settings/billing.js]: @mrsaicharan1 Use RegExp here

File: app/components/forms/admin/settings/billing.js [hunk 20]
```diff
@@ -0,0 +1,133 @@
...
+
+        adminBillingZip: {
+          identifier : 'adminBillingZip',
+          rules      : [
+            {
+              type   : 'empty',
+              prompt : this.l10n.t('Please enter the zip code')
+            },
+            {
+              type   : 'number',
```
Comment (kushthedude) [inline app/components/forms/admin/settings/billing.js]: Remove this `Zip Code` can be alpha numerix

File: app/components/forms/admin/settings/billing.js (hunk 17 above)
Comment (kushthedude) [inline app/components/forms/admin/settings/billing.js]: Empty Line

File: app/components/forms/admin/settings/billing.js [hunk 21]
```diff
@@ -0,0 +1,133 @@
...
+              prompt : this.l10n.t('Please enter the state')
+            }
+          ]
+        },
+        adminBillingAdditionalInfo: {
+          identifier : 'adminBillingAdditionalInfo',
+          rules      : [
+            {
+              type   : 'empty',
+              prompt : this.l10n.t('Please enter additional info')
```
Comment (kushthedude) [inline app/components/forms/admin/settings/billing.js]: Shouldn't additional info be optional?

File: app/routes/admin/settings/billing.js [hunk 22]
```diff
@@ -0,0 +1,7 @@
+import Route from '@ember/routing/route';
+
+export default class extends Route {
+  model() {
+    return this.modelFor('admin.settings');
```
Comment (kushthedude) [inline app/routes/admin/settings/billing.js]: Add `Title Token` .

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 23]
```diff
@@ -0,0 +1,77 @@
...
+    <label for="adminBillingPhone">{{t 'Phone'}}</label>
+    {{input type='text' id='adminBillingPhone' value=model.adminBillingPhone}}
+  </div>
+  <div class="field">
+    <label class="required" for="adminBillingEmail">{{t 'Email'}}</label>
+    {{input type='text' id='adminBillingEmail' value=model.adminBillingEmail}}
+  </div>
+  <div class="field">
+    <label class="required" for="adminBillingCountry">{{t 'Country'}}</label>
+    {{input type='text' id='adminBillingCountry' value=model.adminBillingCountry}}
```
Comment (kushthedude) [inline app/templates/components/forms/admin/settings/billing.hbs]: Make it a country dropdown, We already have a dictionary for this purpose.

File: app/templates/components/forms/admin/settings/billing.hbs [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+  <div class="field">
+    <label class="required" for="adminBillingAddress">{{t 'Address'}}</label>
+    {{textarea rows='2' id='adminBillingAddress' value=model.adminBillingAddress}}
+  </div>
+  <div class="field">
+    <label class="required" for="adminBillingCity">{{t 'City'}}</label>
+    {{input type='text' id='adminBillingCity' value=model.adminBillingCity}}
+  </div>
+  <div class="field">
+    <label class="required" for="adminBillingState">{{t 'State'}}</label>
```
Comment (kushthedude) [inline app/templates/components/forms/admin/settings/billing.hbs]: `State` should be optional

File: app/routes/admin/settings/billing.js (hunk 22 above)
Comment (mrsaicharan1) [reply app/routes/admin/settings/billing.js]: added.

File: app/templates/components/forms/admin/settings/billing.hbs (hunk 24 above)
Comment (kushthedude) [reply app/templates/components/forms/admin/settings/billing.hbs]: @mrsaicharan1 you forgot to remove validations else its good to go

Comment (mrsaicharan1) [reply app/templates/components/forms/admin/settings/billing.hbs]: @kushthedude Done.

**Comment by mrsaicharan1** [comment]
@niranjan94 @kushthedude @uds5501 @shreyanshdwivedi There were some linting errors. Fixed them.

**Comment by mrsaicharan1** [comment]
@CosmicCoder96 I think that we can merge this.

**Pull request merged**